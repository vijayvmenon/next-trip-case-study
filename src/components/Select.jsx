/* eslint-disable */
import React from 'react';
import PropTypes from 'prop-types';
import { createUseStyles } from 'react-jss';
import { colors } from '../constants';
import Dropdown from '../assets/dropdown.svg';

const useStyles = createUseStyles({
  root: {
    position: 'relative',
    marginTop: 24,
  },
  selected: {
    padding: '12px 10px 0px 8px',
    '&:focus': {
      borderColor: colors.BACKGROUND_2,
      borderWidth: 2,
      backgroundPosition: 'top 5px right 3px',
      padding: '12px 10px 0px 7px',
      extend: 'selectFocus',
    },
  },
  notSelected: {
    padding: '2px 10px 0px 5px',
    '&:focus': {
      borderColor: colors.BACKGROUND_2,
      borderWidth: 2,
      backgroundPosition: 'top 5px right 3px',
      padding: '2px 10px 0px 4px',
      extend: 'selectFocus',
    },
  },
  selectFocus: {
    outline: 'none',
    borderRadius: 10,
  },
  select: {
    fontSize: 18,
    borderRadius: 10,
    appearance: 'none',
    backgroundImage: `url(${Dropdown})`,
    background: 'transparent',
    backgroundRepeat: 'no-repeat',
    backgroundPosition: 'top 6px right 4px',
    display: 'block',
    width: '100%',
    color: colors.TEXT_COLOR,
    height: 56,
    border: '1px solid #757575',
    '&:focus ~ label': {
      top: 5,
      left: 9,
      fontSize: '0.8rem',
      color: '#4285f4',
    },
    // "& option[value=''][disabled]": {
    //   display: 'none',
    // },
  },
  label: {
    color: '#999',
    paddingBottom: 10,
    fontSize: '0.8rem',
    fontWeight: 'normal',
    position: 'absolute',
    pointerEvents: 'none',
    left: 9,
    top: 5,
    transition: '0.2s ease all',
  },
});
export default function Select({
  onSearch,
  label,
  labelSelected,
  options = [],
  idField,
  valueField,
  selectedValue,
}) {
  const classes = useStyles();
  return (
    <div className={classes.root}>
      <select
        className={`${classes.select} ${
          selectedValue !== '' ? classes.selected : classes.notSelected
        }`}
        onChange={(e) => {
          onSearch(
            options.find(
              (val) => val[idField].toString() === e.target.value.toString(),
            ),
            labelSelected,
            idField,
          );
        }}
        value={selectedValue}
      >
        <option value={label} disabled hidden>
          {label}
        </option>
        {options.map((val, key) => (
          <option key={val[idField]} value={val[idField]}>
            {val[valueField]}
          </option>
        ))}
      </select>
      {selectedValue !== '' && (
        <label htmlFor={label} className={classes.label}>
          {labelSelected}
        </label>
      )}
    </div>
  );
}

Select.propTypes = {
  onSearch: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
  labelSelected: PropTypes.string.isRequired,
  idField: PropTypes.string.isRequired,
  valueField: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(PropTypes.any),
  selectedValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
};
