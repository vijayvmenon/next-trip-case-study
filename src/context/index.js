import { useContext } from 'react';
import AppContext, { Context } from './context';

export default AppContext;
export { Context };
export const useMyContext = () => useContext(Context);
