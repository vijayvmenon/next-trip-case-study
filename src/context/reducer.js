import constants from '../constants';
import ENTranslations from '../locales/en.json';
import { localStorageExists } from '../utils';

const localStorageLanguage =
  localStorageExists && localStorage.getItem('LANGUAGE');
const initialState = {
  tableExpanded: false,
  hideHeader: false,
  snackbar: { value: '', text: '' },
  language: localStorageLanguage || 'en',
  translations: localStorageLanguage
    ? constants.LANGUAGE_MAPPING[localStorageLanguage]
    : ENTranslations,
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case 'TABLE_EXPAND':
      return { ...state, tableExpanded: true };
    case 'TABLE_COLLAPSE':
      return { ...state, tableExpanded: false };
    case 'SET_SNACKBAR':
      return { ...state, snackbar: { value: action.value, text: action.text } };
    case 'HIDE_HEADER':
      return { ...state, hideHeader: true };
    case 'SHOW_HEADER':
      return { ...state, hideHeader: false };
    case 'SET_LANGUAGE':
      return {
        ...state,
        language: action.value,
        translations: constants.LANGUAGE_MAPPING[action.value],
      };
    default:
      return { ...state };
  }
}

export { initialState };
