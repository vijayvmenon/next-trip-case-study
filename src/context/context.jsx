import React, { useReducer, createContext } from 'react';
import PropTypes from 'prop-types';
import reducer, { initialState } from './reducer';

export const Context = createContext();

export default function AppContext({ children }) {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <Context.Provider value={{ state, dispatch }}>{children}</Context.Provider>
  );
}

AppContext.propTypes = {
  children: PropTypes.objectOf(PropTypes.any).isRequired,
};
