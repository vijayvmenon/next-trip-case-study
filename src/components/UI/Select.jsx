import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { createUseStyles } from 'react-jss';
import { useMediaPredicate } from 'react-media-hook';
import { colors } from '../../constants';
import Dropdown from '../../assets/dropdown.svg';
import { Context } from '../../context';

const useStyles = createUseStyles({
  root: {
    position: 'relative',
    marginTop: 24,
    '@media(max-width:640px)': {
      margin: '24px 3rem 0px 3rem',
    },
  },
  rootLanguage: {
    position: 'relative',
    width: 150,
  },
  selected: {
    padding: '12px 10px 0px 8px',
    '&:focus': {
      borderColor: colors.BACKGROUND_2,
      borderWidth: 2,
      backgroundPosition: 'top 5px right 3px',
      padding: '12px 10px 0px 7px',
      extend: 'selectFocus',
    },
  },
  languageSelected: {
    padding: '0px 10px 0px 8px',
    '&:focus': {
      outline: 'none',
    },
  },
  notSelected: {
    padding: '2px 10px 0px 5px',
    '&:focus': {
      borderColor: colors.BACKGROUND_2,
      borderWidth: 2,
      backgroundPosition: 'top 5px right 3px',
      padding: '2px 10px 0px 4px',
      extend: 'selectFocus',
    },
  },
  selectFocus: {
    outline: 'none',
    borderRadius: 10,
  },
  selectRoot: {
    fontSize: 18,
    borderRadius: 10,
    appearance: 'none',
    backgroundImage: `url(${Dropdown})`,
    background: 'transparent',
    backgroundRepeat: 'no-repeat',
    display: 'block',
    width: '100%',
    color: colors.TEXT_COLOR,

    border: '1px solid #757575',
    '&:focus ~ label': {
      top: 5,
      left: 9,
      fontSize: '0.8rem',
      color: '#4285f4',
    },
  },
  select: {
    extend: 'selectRoot',
    backgroundPosition: 'top 6px right 4px',
    height: 56,
    // "& option[value=''][disabled]": {
    //   display: 'none',
    // },
  },
  languageSelect: {
    extend: 'selectRoot',
    backgroundPosition: 'top 0px right 4px',
    height: 40,
  },
  label: {
    color: '#999',
    paddingBottom: 10,
    fontSize: '0.8rem',
    fontWeight: 'normal',
    position: 'absolute',
    pointerEvents: 'none',
    left: 9,
    top: 5,
    transition: '0.2s ease all',
  },
  selectedOption: {
    background: colors.BACKGROUND_1,
  },
});

export default function Select({
  onSearch,
  label,
  labelSelected,
  options = [],
  idField,
  valueField,
  selectedValue,
  languageDropdown,
}) {
  const classes = useStyles();
  const { state: { translations = {} } = {} } = useContext(Context);
  const normalTable = useMediaPredicate('(min-width: 768px)');

  return (
    <div className={languageDropdown ? classes.rootLanguage : classes.root}>
      <select
        className={`${
          languageDropdown && !normalTable
            ? classes.languageSelect
            : classes.select
        } ${
          // eslint-disable-next-line no-nested-ternary
          selectedValue !== label && languageDropdown && !normalTable
            ? classes.languageSelected
            : selectedValue !== label
            ? classes.selected
            : classes.notSelected
        }`}
        onChange={(e) => {
          onSearch(
            options.find(
              (val) => val[idField].toString() === e.target.value.toString(),
            ),
            labelSelected,
            idField,
          );
        }}
        value={selectedValue}
        data-testid={`custom-select-${idField}`}
      >
        <option
          data-testid={`option-default-${idField}`}
          value={label}
          disabled
          hidden={options.length > 0}
        >
          {translations[label]}
        </option>
        {options.map((val) => (
          <option
            key={val[idField]}
            value={val[idField]}
            className={
              val[idField] === selectedValue
                ? classes.selectedOption
                : undefined
            }
            data-testid={`option-${val[idField]}-${idField}`}
          >
            {val[valueField]}
          </option>
        ))}
      </select>
      {(selectedValue !== label && !languageDropdown) ||
      (languageDropdown && normalTable) ? (
        <label htmlFor={label} className={classes.label}>
          {translations[labelSelected]}
        </label>
      ) : null}
    </div>
  );
}

Select.defaultProps = {
  languageDropdown: false,
};

Select.propTypes = {
  onSearch: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
  labelSelected: PropTypes.string.isRequired,
  idField: PropTypes.string.isRequired,
  valueField: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(PropTypes.any).isRequired,
  selectedValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  languageDropdown: PropTypes.bool,
};
