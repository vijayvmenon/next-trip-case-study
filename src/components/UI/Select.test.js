import { render, fireEvent, screen } from '@testing-library/react';
import Select from './Select';
import renderer from 'react-test-renderer';
import { Context } from '../../context';
import userEvent from '@testing-library/user-event';

const dispatch = jest.fn(); // mock dispatch

const props = {
  onSearch: jest.fn(),
  label: 'test',
  labelSelected: 'test',
  options: [
    {
      test_id: '1',
      test_value: 'value 1',
    },
    {
      test_id: '2',
      test_value: 'value 2',
    },
  ],
  idField: 'test_id',
  valueField: 'test_value',
  selectedValue: '',
  languageDropdown: false,
};

describe('Component:Select.jsx', () => {
  test('Select.jsx renders correctly', () => {
    const select = renderer
      .create(
        <Context.Provider value={{ dispatch }}>
          <Select {...props} />
        </Context.Provider>,
      )
      .toJSON();
    expect(select).toMatchSnapshot();
  });

  test('Select onSearch prop is called on changing select options', () => {
    const { getByTestId, getAllByTestId } = render(
      <Context.Provider value={{ dispatch }}>
        <Select {...props} />
      </Context.Provider>,
    );
    userEvent.selectOptions(getByTestId('custom-select-test_id'), ['value 2']);
    expect(props.onSearch).toHaveBeenCalledTimes(1);
  });
});
