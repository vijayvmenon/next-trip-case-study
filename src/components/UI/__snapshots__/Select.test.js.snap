// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component:Select.jsx Select.jsx renders correctly 1`] = `
<div
  className="root-0-2-1"
>
  <select
    className="select-0-2-8 selected-0-2-3"
    data-testid="custom-select-test_id"
    onChange={[Function]}
    value=""
  >
    <option
      data-testid="option-default-test_id"
      disabled={true}
      hidden={true}
      value="test"
    />
    <option
      data-testid="option-1-test_id"
      value="1"
    >
      value 1
    </option>
    <option
      data-testid="option-2-test_id"
      value="2"
    >
      value 2
    </option>
  </select>
  <label
    className="label-0-2-10"
    htmlFor="test"
  />
</div>
`;
